### PR TITLE
Improved network error handling

### PR DESCRIPTION
**Pull request by tli2**
This is the last network layer refactor PR. I mainly worked on improving error handling using wrappers and exceptions. There is much unfinished work but I should have gotten to most things above protocol handler.

Protocol Handler and traffic cop are both so complicated we probably need a summer rewriting them (if we ever want to)

### CODE REVIEW

**Comment by schedutron** [comment]
This might be lame, but in [error_util.h](src/include/network/error_util.h), wouldn't it be better to write `LogErrno()` like the following?

```
for (int i=0; i < ERRNOS.size(); i++){  //ERRNOS contains EINTR, EAGAIN, EBADF, ...
    if (errno = ERRNOS[i]){
        LOG_TRACE("Error Writing: %s", #ERRNOS[i]);
        break;
    }
}

if (i < ERRNOS.size())
    LOG_TRACE("Error Writing: UNKNOWN");
```

**Comment by tli2** [comment]
This matters little because:

- this is hidden in a function anyways
- this is inherited code from before

I agree that if I were to do it I would do something differently, like keeping a static std::map<int, std::string> from errno to their names. However, a loop does not seem like a good idea since it is first of all slower and second of all non-idiomatic.

File: src/include/network/notifiable_task.h [hunk 1]
```diff
@@ -32,6 +26,7 @@
 #include "common/logger.h"
 #include "container/lock_free_queue.h"
 #include "network_state.h"
+#include "network/error_util.h"
```
Comment (dengyuchenkit) [inline src/include/network/notifiable_task.h]: Can we remove `network/`?

File: src/include/network/notifiable_task.h [hunk 2]
```diff
@@ -59,20 +54,14 @@ class NotifiableTask {
...
-  explicit NotifiableTask(int task_id)
-      : task_id_(task_id) {
-    base_ = event_base_new();
-    // TODO(tianyu) Error handling here can be better perhaps?
-    if (base_ == nullptr) {
-      LOG_ERROR("Can't allocate event base\n");
-      exit(1);
-    }
+  explicit NotifiableTask(int task_id) : task_id_(task_id) {
+    base_ = EventUtil::EventBaseNew();
```
Comment (dengyuchenkit) [inline src/include/network/notifiable_task.h]: Are we handling exceptions when we create `NotifiableTask` like `ConnectionDispatcherTask` or `ConnectionHandlerTask` now?

**Comment by vittvolt** [comment]
@dengyuchenkit One question, does the postgres parser handle exceptions well and send through the network thread?

**Comment by dengyuchenkit** [comment]
@vittvolt The network layer will catch the exception thrown by the parser and send back the error message. The parser detects different exceptions now but I am not sure what do you mean by `parser handle exception well`? Are you talking about particular exceptions that should be detected by parser? If it's only about network thread catching exception, then it's yes.

File: src/include/network/notifiable_task.h (hunk 2 above)
Comment (tli2) [reply src/include/network/notifiable_task.h]: They should both only interact with libevent through the NotifiableTask interface. (the base is private and not accessible to child classes)

**Comment by vittvolt** [comment]
@dengyuchenkit Thanks for the clarification. The problem I found was that exceptions thrown in postgresparser cause peloton the terminate, though the exception info is printed.

Not sure if it is related to this PR.

**Comment by dengyuchenkit** [comment]
@vittvolt >< I think this PR is more about refactoring. Can you please share the queries you run? Will check that...

File: src/include/network/notifiable_task.h (hunk 1 above)
Comment (tli2) [reply src/include/network/notifiable_task.h]: Will do.

**Comment by dengyuchenkit** [comment]
@vittvolt Sorry, some corrections. This PR is removing `EXIT`. And I'm not sure about the terminate you mention. Could you share the queries?

**Comment by apavlo** [comment]
@vittvolt We should catch those and send them back to the client.

@dengyuchenkit This PR is separate from your work on passing exceptions from the worker threads to the networking thread, right?

File: src/include/network/error_util.h [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+namespace peloton {
+namespace network {
+/**
+ * Static utility class for some common error handling code
+ */
+class ErrorUtil {
+public:
+  ErrorUtil() = delete;
+
+  static void LogErrno() {
```
Comment (apavlo) [inline src/include/network/error_util.h]: What is this function? And why are we using `LOG_TRACE` instead of `LOG_ERROR`.

**Comment by dengyuchenkit** [comment]
@apavlo Yes, it's separate.

File: src/include/network/notifiable_task.h [hunk 4]
```diff
@@ -171,29 +160,21 @@ class NotifiableTask {
...
-//      LOG_ERROR("Failed to update event");
-//      PL_ASSERT(false);
-//    }
-//    event_add(event, timeout);
-//  }
-//  void UpdateManualEvent(struct event *event, event_callback_fn callback, void *arg) {
-//    UpdateEvent(event, -1, EV_PERSIST, callback, arg);
-//  }
+// TODO(tianyu): The original network code seems to do this as an optimization. I am leaving this out until we get numbers
+  void UpdateEvent(struct event *event,
```
Comment (apavlo) [inline src/include/network/notifiable_task.h]: Can we add documentation for all of these parameters?

File: src/network/connection_handle.cpp [hunk 5]
```diff
@@ -448,36 +374,33 @@ bool ConnectionHandle::ProcessInitialPacket(InputPacket *pkt) {
...
-    LOG_ERROR("Protocol error: Only protocol version 3 is supported.");
-    exit(EXIT_FAILURE);
-  }
+  if (PROTO_MAJOR_VERSION(proto_version) != 3)
+    throw NetworkProcessException("Protocol error: Only protocol version 3 is supported.");
 
   // TODO: check for more malformed cases
   // iterate till the end
+  // TODO(tianyu): WTF is this?
   for (;;) {
```
Comment (apavlo) [inline src/network/connection_handle.cpp]: Yeah seriously?? What is this???

File: src/include/network/error_util.h (hunk 3 above)
Comment (tli2) [reply src/include/network/error_util.h]: This is migrated from clearly copy-and-pasted errno handling in the protocol handler. I suppose I can make it better as well.

File: src/include/network/notifiable_task.h (hunk 4 above)
Comment (tli2) [reply src/include/network/notifiable_task.h]: Yes

**Pull request merged**